### PR TITLE
Add API for get the memory stats

### PR DESCRIPTION
**Pull request by jiangzidong**
During our developing runtime based on Jerry, we find that it is needed to get the heap stats inside jerry engine. e.g. to impl `performance.memory` 
https://developer.mozilla.org/en-US/docs/Web/API/Window/performance
https://anonyco.github.io/wiki/apis/timing/properties/memory/
```
console.log(performance.memory)

// Would show, for example
//{
// jsHeapSizeLimit: 767557632,
// totalJSHeapSize: 58054528,
// usedJSHeapSize: 42930044
//}
```
In this PR, I added `jerry_get_memory_stats` in API.

And during test, I found the current `jmem_heap_stats_print` have `divided by 0` issue (block test-heap), and I quickly fixed it (not a decent way)
 
JerryScript-DCO-1.0-Signed-off-by: Zidong Jiang zidong.jiang@intel.com

### CODE REVIEW

**Comment by jiangzidong** [comment]
cc @martijnthe

File: docs/02.API-REFERENCE.md [hunk 1]
```diff
@@ -169,6 +169,51 @@ typedef struct
...
+  size_t peak_property_bytes; /**< peak allocated memory for properties */
+
+  size_t skip_count; /**< Number of skip-aheads during insertion of free block */
+  size_t nonskip_count; /**< Number of times we could not skip ahead during
+                         *   free block insertion */
+
+  size_t alloc_count; /**< number of memory allocations */
+  size_t free_count; /**< number of memory frees */
+  size_t alloc_iter_count; /**< Number of iterations required for allocations */
+  size_t free_iter_count; /**< Number of iterations required for inserting free blocks */
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: I would really not make this structure public. It contains implementation details which might change.

Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: I agree

**Comment by zherczeg** [comment]
We could have a public structure with version number to publish this data. So we could extend the structure later without breaking compatibility.

**Comment by jiangzidong** [comment]
@zherczeg @LaszloLango What about exposing the following struct?
```c
typedef struct
{
  size_t size; /**< heap total size */
  size_t allocated_bytes; /**< currently allocated bytes */
  size_t peak_allocated_bytes; /**< peak allocated bytes */
  uint8_t version;
} jerry_heap_stats_t
```
Because in the current stage, we are mainly interested in those members.

**Comment by zherczeg** [comment]
Something like this. As for version number, it can be part of the struct, or passed directly to the function. I would prefer the latter since it is more robust. Anyway, if it is part of the struct, it must be the first member.

**Comment by jiangzidong** [comment]
Thanks, but I think the function call like `jerry_get_memory_stats(out_stats_p, 1)`, the version number here is a bit weird, because the caller need to know exactly the version of the mem-stats struct.

I will firstly impl the "version inside struct" way.

File: docs/02.API-REFERENCE.md [hunk 2]
```diff
@@ -523,6 +544,40 @@ main (void)
...
+
+- `out_stats_p` - out parameter, that gives the stats of the heap.
+- return value
+  - true, if run was successful
+  - false, otherwise. Usually it is because the MEM_STATS feature is not enabled.
+
+**Example**
+
+```c
+jerry_init (JERRY_INIT_EMPTY | JERRY_INIT_MEM_STATS);
```
Comment (martijnthe) [inline docs/02.API-REFERENCE.md]: I was trying to find how `JERRY_INIT_MEM_STATS` affects the engine but grepping for this define didn't lead to much insight. Can someone point me to how this flag is actually being consumed?

File: jerry-core/api/jerry.c [hunk 3]
```diff
@@ -254,6 +255,34 @@ jerry_gc (void)
...
+ * @return true - get the heap stats successful
+ *         false - otherwise. Usually it is because the MEM_STATS feature is not enabled.
+ */
+bool
+jerry_get_memory_stats (jerry_heap_stats_t *out_stats_p) /**< [out] heap memory stats */
+{
+#ifdef JMEM_STATS
+  jmem_heap_stats_t jmem_heap_stats = {0};
+  jmem_heap_get_stats (&jmem_heap_stats);
+
```
Comment (martijnthe) [inline jerry-core/api/jerry.c]: check for `NULL == out_stats_p` and return `false`?

File: jerry-core/api/jerry.c [hunk 4]
```diff
@@ -254,6 +255,34 @@ jerry_gc (void)
...
+  out_stats_p->version = 1;
```
Comment (martijnthe) [inline jerry-core/api/jerry.c]: I would make sure to completely initialize the struct.
Personally I like doing this as such:

```
*out_stats_p = (jerry_heap_stats_t) {
  .size-> = ...,
 // ...
 // if more fields are added to the struct in the future, but we forget them here,
 // they will at least be initialized to zero (instead of being garbage).
};
```

**Comment by fbmrk** [comment]
@jiangzidong 
Possibly the `waste_bytes` also could be useful information, for example before a `gc()` call.
What do you think?

File: jerry-core/include/jerryscript-core.h [hunk 5]
```diff
@@ -159,6 +159,18 @@ typedef struct
 } jerry_property_descriptor_t;
 
 /**
+ * Description of JerryScript heap memory stats.
+ * It is for memory profiling.
+ */
+typedef struct
+{
+  size_t version; /**< the version of the stats struct */
```
Comment (martijnthe) [inline jerry-core/include/jerryscript-core.h]: I'm not really convinced `version` is needed.
IMO, the struct and field types and names are a semantic API contract, but IMO the struct layout isn't an ABI (the exact layout is target- & compiler dependent after all...)
I think it's safe to assume that the users of JerryScript have control over what version of Jerry they use.
If the semantics change (i.e. field is renamed, types are changed, ...) the compiler will tell you.

Comment (jiangzidong) [inline jerry-core/include/jerryscript-core.h]: @zherczeg @LaszloLango @fbmrk @martijnthe 
What about 
```
typedef struct
{
  size_t size; /**< heap total size */
  size_t allocated_bytes; /**< currently allocated bytes */
  size_t peak_allocated_bytes; /**< peak allocated bytes */
  size_t waste_bytes; /**< bytes waste due to blocks filled partially */
  size_t peak_waste_bytes;
}  jerry_heap_stats_t
```
I think those fields are not much impl-dependent comparing to `alloc_count`, `alloc_iter_count`.
And maybe we can add `byte_code_bytes`, `string_bytes`, and so on.

File: docs/02.API-REFERENCE.md (hunk 2 above)
Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: in jmem-allocator.c
```c
void
jmem_finalize (void)
{
  jmem_pools_finalize ();

#ifdef JMEM_STATS
  if (JERRY_CONTEXT (jerry_init_flags) & ECMA_INIT_MEM_STATS)
  {
    jmem_stats_print ();
  }
#endif /* JMEM_STATS */

  jmem_heap_finalize ();
} /* jmem_finalize */
```
I think the only usage of "JERRY_INIT_MEM_STATS" is dumping the mem stats after cleaning up. So you're right, we don't need it here, i will remove it.

File: jerry-core/include/jerryscript-core.h (hunk 5 above)
Comment (zherczeg) [inline jerry-core/include/jerryscript-core.h]: The problem is if you dynamically link Jerry, and you compiled your code with an older Jerry, and the system has a newer one, you will experience a crash. Currently most projects statically link Jerry, so this is not a problem, but it might be if some distros will support it.

**Comment by jiangzidong** [comment]
@zherczeg @fbmrk @martijnthe 
updated the patch

File: jerry-core/include/jerryscript-core.h (hunk 5 above)
Comment (martijnthe) [inline jerry-core/include/jerryscript-core.h]: > The problem is if you dynamically link Jerry, and you compiled your code with an older Jerry, and the system has a newer one, you will experience a crash.

In the current implementation doesn't know or check the version / how large the caller's `jerry_heap_stats_t` struct is. I think the caller should be passing in the desired/expected version then for this to work?

Comment (martijnthe) [inline jerry-core/include/jerryscript-core.h]: @zherczeg I don't know how versioning of APIs with dynamic libraries this is usually dealt with, but I wonder whether it makes sense to have a over-arching "get version" API for version matching instead of adding versions for each and every API / ABI.

Comment (zherczeg) [inline jerry-core/include/jerryscript-core.h]: But you don't know the changes in future versions. At compile time you see the correct version, but when dynamically linked, you get a library with different version.

File: docs/02.API-REFERENCE.md [hunk 6]
```diff
@@ -523,6 +546,36 @@ main (void)
...
+
+- `out_stats_p` - out parameter, that gives the stats of the heap.
+- return value
+  - true, if run was successful
+  - false, otherwise. Usually it is because the MEM_STATS feature is not enabled.
+
+**Example**
+
+```c
+jerry_heap_stats_t stats = {0};
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: I would prefer a textual version number.

File: docs/02.API-REFERENCE.md [hunk 7]
```diff
@@ -169,6 +169,29 @@ typedef struct
...
+typedef struct
+{
+  size_t version /**< the version of the stats struct */
+  size_t size; /**< heap total size */
+  size_t allocated_bytes; /**< currently allocated bytes */
+  size_t peak_allocated_bytes; /**< peak allocated bytes */
+  size_t waste_bytes; /**< bytes waste due to blocks filled partially */
+  size_t peak_waste_bytes; /**< peak wasted bytes */
+} jerry_heap_stats_t;
+
```
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Unnecessary empty line

File: docs/02.API-REFERENCE.md (hunk 6 above)
Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: Sorry, a little confused. Do you mean the `version` is a input param and  `jerry_get_mem_stats` should check the `version`? But in this way, user must know the version clearly and explicitly. I doubt if it is a  good way.

File: jerry-core/api/jerry.c [hunk 8]
```diff
@@ -254,6 +255,44 @@ jerry_gc (void)
   ecma_gc_run (JMEM_FREE_UNUSED_MEMORY_SEVERITY_LOW);
 } /* jerry_gc */
 
+#ifdef JMEM_STATS
+void jmem_heap_get_stats (jmem_heap_stats_t *);
```
Comment (LaszloLango) [inline jerry-core/api/jerry.c]: Why is this deceleration needed here?

File: jerry-core/include/jerryscript-core.h (hunk 5 above)
Comment (LaszloLango) [inline jerry-core/include/jerryscript-core.h]: The heap size is set at compile time. Not sure it is a useful information.

Comment (martijnthe) [inline jerry-core/include/jerryscript-core.h]: > But you don't know the changes in future versions. At compile time you see the correct version, but when dynamically linked, you get a library with different version.
 
Sure, I understand.

> The heap size is set at compile time. Not sure it is a useful information.

Not if you use `jerry_create_instance`.

File: jerry-core/api/jerry.c [hunk 9]
```diff
@@ -254,6 +255,44 @@ jerry_gc (void)
...
+{
+#ifdef JMEM_STATS
+  if (out_stats_p == NULL)
+  {
+    return false;
+  }
+
+  jmem_heap_stats_t jmem_heap_stats = {0};
+  jmem_heap_get_stats (&jmem_heap_stats);
+
```
Comment (martijnthe) [inline jerry-core/api/jerry.c]: Thinking this further:

Say today `sizeof(jerry_heap_stats_t) == 24`.
Say I dynamically link my project against JerryScript of today, so my project does something like:
```
jerry_heap_stats_t stats = {}; // 24 bytes allocated on the stack
jerry_get_memory_stats(&stats); // 24 bytes written
```

Now tomorrow I have a newer JerryScript, which happened to have added some fields to `jerry_heap_stats_t`, so now `sizeof(jerry_heap_stats_t) == 36`.

Now it will be a problem if the project dynamically links to the newer JerryScript:
```
jerry_heap_stats_t stats = {}; // still 24 bytes allocated on the stack
jerry_get_memory_stats(&stats); // 36 bytes written -- WHOOPS, smashing the stack!
```

That's what I meant with "I think the caller should be passing in the desired/expected version then for this to work" in [my earlier comment].

cc @zherczeg 

[my earlier comment]: https://github.com/jerryscript-project/jerryscript/pull/1923#discussion_r128181423

File: jerry-core/api/jerry.c (hunk 8 above)
Comment (jiangzidong) [inline jerry-core/api/jerry.c]: or should I use
```
#define JMEM_ALLOCATOR_INTERNAL
#include "jmem-allocator-internal.h"
```

File: jerry-core/api/jerry.c (hunk 9 above)
Comment (zherczeg) [inline jerry-core/api/jerry.c]: We might talk about the same thing.

    if (jerry_get_memory_stats(VERSION, &stats))

This work regardless of future changes. Each caller gets the data in their expected format. I do think memory statistics is something I expect to change in the future.

Comment (jiangzidong) [inline jerry-core/api/jerry.c]: @zherczeg is that means the jerry_get_memory_stats will check the version, if it is not match, then return false ?

Comment (zherczeg) [inline jerry-core/api/jerry.c]: Not necessarily. It can support multiple variants since the version determines the request. So the implementation will be a switch-case which fills the appropriate fields, or false if no mem-stats or version is unknown.

Comment (jiangzidong) [inline jerry-core/api/jerry.c]: Ok, `if (jerry_get_memory_stats(VERSION, &stats))` the VERSION here should be sth like `JERRY_API_MAJOR_VERSION`. I still feel strange because it is the only API who pass the api version. 
e.g.
`jerry_get_own_property_descriptor` is also getting a block of memory as jerry_property_descriptor_t, but it doesnt pass the api version.

Comment (zherczeg) [inline jerry-core/api/jerry.c]: True, but I feel this is something which will likely change in the future. I have an alternative idea: the version can be encoded in the function name: `jerry_get_memory_stats_1` and people can choose the appropriate one. The compiler will only link the appropriate one.

Comment (martijnthe) [inline jerry-core/api/jerry.c]: @zherczeg Personally, I was also surprised about your concerns w.r.t. version mismatches when dynamic linking, because – as @jiangzidong also pointed out – the rest of the API is not really taking that concern into account at all. Breaking ABI changes have been made in the past without anyone complaining that their dynamically linked libraries became incompatible.

I also still feel it makes the API pretty ugly and heavy if you have to pass in versions all over the place. I'd rather have a over-arching "JerryScript version" that the caller can use to check for compatibility.

Comment (zherczeg) [inline jerry-core/api/jerry.c]: Normally I don't see the reason for versioning, because most functions are simple enough and will likely not change (e.g. I don't think an `is_boolean` will ever change). But I feel memory statistics (as the name suggests) is different. It would be great to have at least a plan for what will happen then.

Comment (martijnthe) [inline jerry-core/api/jerry.c]: Another idea:
Add a bit of padding to the struct for future use, for example something like:
```
typedef struct
{
  int version;
  size_t size; /**< heap total size */
  size_t allocated_bytes; /**< currently allocated bytes */
  size_t peak_allocated_bytes; /**< peak allocated bytes */
  uint32_t reserved[4];
}  jerry_heap_stats_t
```

It won't be extensible indefinitely of course, but it's simple and wouldn't require passing in the version in the call. We'd promise not to grow the struct ever. The caller can check the version after the call.

Comment (zherczeg) [inline jerry-core/api/jerry.c]: Sounds reasonable. If we ever need a much bigger structure, we can create a new function for it. Please make the type of reserved as `size_t`, not `uint32_t`.

**Comment by jiangzidong** [comment]
Please review again, @zherczeg @LaszloLango @martijnthe

File: jerry-core/api/jerry.c (hunk 9 above)
Comment (LaszloLango) [inline jerry-core/api/jerry.c]: I don't like the reserved fields. But this is only my personal preference. My experience is that they are always remaining intact. This is not an objection. If everyone vote for reserved fields than I can accept that. 

> But I feel memory statistics (as the name suggests) is different. It would be great to have at least a plan for what will happen then.

I think this won't change frequently. @zherczeg do you have any plan with memory statistics or management in the near future?

**Comment by martijnthe** [comment]
lgtm

**Comment by martijnthe** [comment]
@LaszloLango could you please take another look? Thanks!

File: docs/02.API-REFERENCE.md [hunk 10]
```diff
@@ -169,6 +169,29 @@ typedef struct
...
+{
+  size_t version /**< the version of the stats struct */
+  size_t size; /**< heap total size */
+  size_t allocated_bytes; /**< currently allocated bytes */
+  size_t peak_allocated_bytes; /**< peak allocated bytes */
+  size_t waste_bytes; /**< bytes waste due to blocks filled partially */
+  size_t peak_waste_bytes; /**< peak wasted bytes */
+  size_t reserved[4]; /**< the padding for future extension */
+} jerry_heap_stats_t;
+
```
Comment (LaszloLango) [inline docs/02.API-REFERENCE.md]: Unnecessary empty line

File: docs/02.API-REFERENCE.md [hunk 11]
```diff
@@ -169,6 +169,28 @@ typedef struct
...
+```c
+typedef struct
+{
+  size_t version /**< the version of the stats struct */
+  size_t size; /**< heap total size */
+  size_t allocated_bytes; /**< currently allocated bytes */
+  size_t peak_allocated_bytes; /**< peak allocated bytes */
+  size_t waste_bytes; /**< bytes waste due to blocks filled partially */
+  size_t peak_waste_bytes; /**< peak wasted bytes */
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: Do we need wasted data? It seems allocator specific, and we cannot provide for normal allocators.

Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: It's fbmrk's idea. What do you think? @fbmrk 
Personally, I also agree that the `waste` is much more impl-specific than `allocated` and `total`

Comment (fbmrk) [inline docs/02.API-REFERENCE.md]: It was just a suggestion, i do not insist on it.

Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: Ok, I will remove it in this patch. If someday we really need it, let's put it into another PR.

Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: @zherczeg update, please review again @fbmrk @zherczeg

File: docs/02.API-REFERENCE.md [hunk 12]
```diff
@@ -169,6 +169,26 @@ typedef struct
...
+**Prototype**
+
+```c
+typedef struct
+{
+  size_t version /**< the version of the stats struct */
+  size_t size; /**< heap total size */
+  size_t allocated_bytes; /**< currently allocated bytes */
+  size_t peak_allocated_bytes; /**< peak allocated bytes */
+  size_t reserved[4]; /**< the padding for future extension */
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: padding for future extensions

File: docs/02.API-REFERENCE.md [hunk 13]
```diff
@@ -523,6 +543,36 @@ main (void)
...
+Get heap memory stats.
+
+**Prototype**
+
+```c
+bool
+jerry_get_memory_stats (jerry_heap_stats_t *out_stats_p);
+```
+
+- `out_stats_p` - out parameter, that gives the stats of the heap.
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: that provides the heap statistics

File: jerry-core/api/jerry.c [hunk 14]
```diff
@@ -33,9 +33,14 @@
 #include "ecma-promise-object.h"
 #include "jcontext.h"
 #include "jerryscript.h"
+#include "jmem.h"
 #include "js-parser.h"
 #include "re-compiler.h"
 
+#define JMEM_ALLOCATOR_INTERNAL
+#include "jmem-allocator-internal.h"
+#undef JMEM_ALLOCATOR_INTERNAL
```
Comment (zherczeg) [inline jerry-core/api/jerry.c]: Why do we need this include? If something is needed, we should move it to jmem.h.

File: jerry-core/jmem/jmem-heap.c [hunk 15]
```diff
@@ -615,12 +615,12 @@ jmem_heap_stats_print (void)
...
-                   heap_stats->alloc_iter_count / heap_stats->alloc_count,
-                   heap_stats->alloc_iter_count % heap_stats->alloc_count * 10000 / heap_stats->alloc_count,
-                   heap_stats->free_iter_count / heap_stats->free_count,
-                   heap_stats->free_iter_count % heap_stats->free_count * 10000 / heap_stats->free_count);
+                   heap_stats->skip_count / (heap_stats->nonskip_count + 1),
+                   heap_stats->skip_count % (heap_stats->nonskip_count + 1) * 10000 / (heap_stats->nonskip_count + 1),
+                   heap_stats->alloc_iter_count / (heap_stats->alloc_count + 1),
+                   heap_stats->alloc_iter_count % (heap_stats->alloc_count + 1) * 10000 / (heap_stats->alloc_count + 1),
+                   heap_stats->free_iter_count / (heap_stats->free_count + 1),
+                   heap_stats->free_iter_count % (heap_stats->free_count + 1) * 10000 / (heap_stats->free_count + 1));
```
Comment (zherczeg) [inline jerry-core/jmem/jmem-heap.c]: I suspect the +1 is needed for avoiding division by zero, but the formula is not correct for non-zero values. I would introduce a helper function for this. Anyway this change seems unrelated, and perhaps we could fix it in another patch.

Comment (jiangzidong) [inline jerry-core/jmem/jmem-heap.c]: yes, to avoid the "divide by 0" exception

Comment (zherczeg) [inline jerry-core/jmem/jmem-heap.c]: What about fixing it in another patch?

Comment (jiangzidong) [inline jerry-core/jmem/jmem-heap.c]: because the unittest `test-poolmain,c` calls `jmem_heap_stats_print`, which will cause error if we don't fix it. (JMEM_STATS is not enabled before, so the issue is not be discovered)

What about comments out the function call, and I will reset the change for function body.

Comment (zherczeg) [inline jerry-core/jmem/jmem-heap.c]: Hm, ok. Lets disable it for now (leave a TODO comment). This issue may affect other things, so it deserves its own patch.

Comment (jiangzidong) [inline jerry-core/jmem/jmem-heap.c]: Ok, updated, please review it again. Thanks
